go/arrow/flight/flightsql/client.go

```diff
@@ -218,6 +218,58 @@ func (c *Client) ExecuteSubstraitUpdate(ctx context.Context, plan SubstraitPlan,
 	return updateResult.GetRecordCount(), nil
 }
 
+var ErrTooManyPutResults = fmt.Errorf("%w: server sent multiple PutResults, expected one ", arrow.ErrInvalid)
+
+// ExecuteIngest is for executing a bulk ingestion and only returns the number of affected rows.
+func (c *Client) ExecuteIngest(ctx context.Context, rdr array.RecordReader, reqOptions *ExecuteIngestOpts, opts ...grpc.CallOption) (n int64, err error) {
+	var (
+		desc         *flight.FlightDescriptor
+		stream       pb.FlightService_DoPutClient
+		wr           *flight.Writer
+		res          *pb.PutResult
+		updateResult pb.DoPutUpdateResult
+	)
+
+	cmd := (*pb.CommandStatementIngest)(reqOptions)
+	if desc, err = descForCommand(cmd); err != nil {
+		return
+	}
+
+	if stream, err = c.Client.DoPut(ctx, opts...); err != nil {
+		return
+	}
+
+	wr = flight.NewRecordWriter(stream, ipc.WithAllocator(c.Alloc), ipc.WithSchema(rdr.Schema()))
+	defer wr.Close()
+
+	wr.SetFlightDescriptor(desc)
+
+	for rdr.Next() {
+		rec := rdr.Record()
+		wr.Write(rec)
+	}
+
```

Comment:
there's some error handling to be done here...

We should check the return of `wr.Write(rec)` which returns an `error`, in case we encounter an error while writing.

In addition, after this loop we should probably check `rdr.Err()` in case the reader itself encountered an error.

Also, is this function responsible for calling `Release` on the reader? or is the calling function responsible?

Reply:
I updated this to handle the error cases you described.

I looked at a few places in the code where a function consumes a RecordReader to see the `Release` semantics they use. The pattern I generally saw was that the callee retains the reader while using it and then releases it when finished, but it is still the caller's responsibility to release the original reference to the reader. I went with those same semantics here and updated the godoc comment to indicate that. I enabled allocation checking in the tests to ensure these kinds of things get caught going forward.

I also added some client-side validation of the request pb for fields that are not considered optional in the proto spec.

Reply:
Just a heads up: I expanded the testing to check some trickier scenarios like if the server returned an error mid-DoPut stream. Between this and handling some other flaky test failures, the error handling has been changed/improved quite a bit again. Just in case you wanted to review again.